### PR TITLE
server: fix expiry detection for verification keys

### PR DESCRIPTION
**Pull request by ericchiang**
updates #828

### CODE REVIEW

File: server/rotation.go [hunk 1]
```diff
@@ -132,7 +132,7 @@ func (k keyRotater) rotate() error {
 		i := 0
 
 		for _, key := range keys.VerificationKeys {
-			if !key.Expiry.After(tNow) {
```
Comment (dghubble) [inline server/rotation.go]: // key is still valid, include it in the VerificationKeys

File: server/rotation.go [hunk 2]
```diff
@@ -132,7 +132,7 @@ func (k keyRotater) rotate() error {
...
+			if key.Expiry.After(tNow) {
```
Comment (dghubble) [inline server/rotation.go]: // key is still valid, include it in the VerificationKeys

**Comment by ericchiang** [comment]
@rithujohn191 can you take a look at the PR again? I've added a few more comments to the rotation code.

File: server/rotation.go [hunk 3]
```diff
@@ -128,22 +129,30 @@ func (k keyRotater) rotate() error {
...
-		i := 0
+		expired := func(key storage.VerificationKey) bool {
+			return tNow.After(key.Expiry)
+		}
 
+		// Remove any verification keys that have expired.
+		i := 0
 		for _, key := range keys.VerificationKeys {
-			if !key.Expiry.After(tNow) {
+			if !expired(key) {
```
Comment (ericchiang) [inline server/rotation.go]: Only functional change. Everything else is comments and renaming.

File: server/rotation.go [hunk 4]
```diff
@@ -128,22 +129,30 @@ func (k keyRotater) rotate() error {
...
-			// Move current signing key to a verification only key.
+			// Move current signing key to a verification only key, throwing
+			// away the private part.
 			verificationKey := storage.VerificationKey{
 				PublicKey: keys.SigningKeyPub,
-				Expiry:    tNow.Add(k.strategy.verifyFor),
+				// After demoting the signing key, keep the token around for at least
+				// the amount of time an ID Token is valid for. This ensures the
+				// verification key won't expire until all ID Tokens is could have
+				// signed have expired as well.
```
Comment (rithujohn191) [inline server/rotation.go]: "This ensures the verification key won't expire until all ID Tokens it has signed have expired as well"?

Comment (ericchiang) [inline server/rotation.go]: thanks, updated.

**Pull request merged**